File: typescript/test/client.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { API_VERSION } from "../src/apiVersion";
import {
  ApiException,
  ErrorCode,
  Meteroid,
  type MeteroidOptions,
  OAuthErrorCode,
} from "../src/index";
import { InvoiceStatus } from "../src/models/invoiceStatus";

const BASE_URL = "https://mock.meteroid.test";

const PAGINATION = { page: 0, per_page: 10, total_items: 0, total_pages: 0 };

type Call = { url: URL; init: RequestInit & { headers: Record<string, string> } };

/**
 * Build a client whose `fetch` is a recorder returning the canned responses, one
 * per call (the last one is reused once the list runs out).
 */
function mockClient(
  responses: Array<{ status: number; body?: unknown }>,
  options: Partial<MeteroidOptions> = {}
): { client: Meteroid; calls: Call[] } {
  const calls: Call[] = [];
  let index = 0;

  const fetchImpl: typeof fetch = async (input, init) => {
    calls.push({
      url: new URL(String(input)),
      init: (init ?? {}) as Call["init"],
    });
    const canned = responses[Math.min(index, responses.length - 1)];
    index += 1;
    return new Response(canned.body === undefined ? "" : JSON.stringify(canned.body), {
      status: canned.status,
      headers: { "content-type": "application/json" },
    });
  };

  const client = new Meteroid("test-api-key", {
    serverUrl: BASE_URL,
    fetch: fetchImpl,
    ...options,
  } as MeteroidOptions);

  return { client, calls };
}

describe("request wiring", () => {
  it("sends the bearer token and the standard headers", async () => {
    const { client, calls } = mockClient([
      { status: 200, body: { data: [], pagination_meta: PAGINATION } },
    ]);

    await client.invoices.listInvoices();

    assert.equal(calls.length, 1);
    assert.equal(calls[0].url.pathname, "/api/v1/invoices");
    assert.equal(calls[0].init.headers.authorization, "Bearer test-api-key");
    assert.match(calls[0].init.headers["user-agent"], /^meteroid-typescript\//);
    assert.strictEqual(calls[0].init.headers["meteroid-version"], API_VERSION);
  });

  it("explodes array query parameters into repeated keys", async () => {
    const { client, calls } = mockClient([
      { status: 200, body: { data: [], pagination_meta: PAGINATION } },
    ]);

    await client.invoices.listInvoices({
      statuses: [InvoiceStatus.Draft, InvoiceStatus.Finalized],
      page: 2,
    });

    const { searchParams } = calls[0].url;
    assert.deepEqual(searchParams.getAll("statuses"), ["DRAFT", "FINALIZED"]);
    assert.equal(searchParams.get("page"), "2");
    assert.equal(calls[0].url.search, "?statuses=DRAFT&statuses=FINALIZED&page=2");
  });

  it("attaches an auto idempotency key to POSTs only", async () => {
    const { client, calls } = mockClient([
      { status: 200, body: { data: [], pagination_meta: PAGINATION } },
    ]);

    await client.invoices.listInvoices();

    assert.equal(calls[0].init.headers["idempotency-key"], undefined);
  });

  it("logs one line per request and response when debug is on", async () => {
    const { client } = mockClient(
      [{ status: 200, body: { data: [], pagination_meta: PAGINATION } }],
      { debug: true }
    );

    const lines: string[] = [];
    const original = console.error;
    console.error = (...args: unknown[]) => {
      lines.push(args.map(String).join(" "));
    };
    try {
      await client.invoices.listInvoices();
    } finally {
      console.error = original;
    }

    assert.equal(lines.length, 2);
    assert.match(lines[0], /^meteroid: GET .*\/api\/v1\/invoices \(attempt 1\)$/);
    assert.match(lines[1], /^meteroid: GET .*\/api\/v1\/invoices -> 200$/);
  });

  it("logs nothing when debug is off", async () => {
    const { client } = mockClient([
      { status: 200, body: { data: [], pagination_meta: PAGINATION } },
    ]);

    const lines: string[] = [];
    const original = console.error;
    console.error = (...args: unknown[]) => {
      lines.push(args.map(String).join(" "));
    };
    try {
      await client.invoices.listInvoices();
    } finally {
      console.error = original;
    }

    assert.equal(lines.length, 0);
  });

  it("parses a RestErrorResponse body", async () => {
    const { client } = mockClient([
      { status: 404, body: { code: "NOT_FOUND", message: "no such customer" } },
    ]);

    const err = await client.customers
      .getCustomer("nope")
      .then(() => null)
      .catch((e: unknown) => e);

    assert.ok(err instanceof ApiException);
    assert.equal(err.status, 404);
    assert.deepEqual(err.restError, {
      code: ErrorCode.NotFound,
      message: "no such customer",
    });
    assert.equal(err.oauthError, undefined);
    assert.equal(err.body, '{"code":"NOT_FOUND","message":"no such customer"}');
    assert.equal(err.message, "API Error 404: NOT_FOUND: no such customer");
  });

  it("parses an OAuthErrorResponse body", async () => {
    const { client } = mockClient([{ status: 400, body: { error: "invalid_grant" } }]);

    const err = await client.invoices
      .listInvoices()
      .then(() => null)
      .catch((e: unknown) => e);

    assert.ok(err instanceof ApiException);
    assert.equal(err.status, 400);
    assert.equal(err.restError, undefined);
    assert.equal(err.oauthError?.error, OAuthErrorCode.InvalidGrant);
    assert.equal(err.oauthError?.errorDescription, undefined);
    assert.equal(err.body, '{"error":"invalid_grant"}');
  });

  it("handles a 422 like any other status", async () => {
    const { client } = mockClient([
      { status: 422, body: { code: "BAD_REQUEST", message: "invalid page" } },
    ]);

    const err = await client.invoices
      .listInvoices()
      .then(() => null)
      .catch((e: unknown) => e);

    assert.ok(err instanceof ApiException);
    assert.equal(err.status, 422);
    assert.equal(err.restError?.code, ErrorCode.BadRequest);
    assert.equal(err.restError?.message, "invalid page");
  });

  it("leaves the payload untyped for an unknown error code", async () => {
    const { client } = mockClient([
      { status: 409, body: { code: "SOMETHING_NEW", message: "later API version" } },
    ]);

    const err = await client.invoices
      .listInvoices()
      .then(() => null)
      .catch((e: unknown) => e);

    assert.ok(err instanceof ApiException);
    assert.equal(err.status, 409);
    assert.equal(err.restError, undefined);
    assert.equal(err.oauthError, undefined);
    assert.equal(err.body, '{"code":"SOMETHING_NEW","message":"later API version"}');
  });

  it("keeps the status and raw body when an error body is not JSON", async () => {
    const fetchImpl: typeof fetch = async () =>
      new Response("<html>gateway error</html>", {
        status: 400,
        headers: { "x-request-id": "abc" },
      });
    const client = new Meteroid("test-api-key", {
      serverUrl: BASE_URL,
      fetch: fetchImpl,
    });

    const err = await client.invoices
      .listInvoices()
      .then(() => null)
      .catch((e: unknown) => e);

    assert.ok(err instanceof ApiException);
    assert.equal(err.status, 400);
    assert.equal(err.body, "<html>gateway error</html>");
    assert.equal(err.headers["x-request-id"], "abc");
    assert.equal(err.restError, undefined);
    assert.equal(err.oauthError, undefined);
    assert.equal(err.message, "API Error 400");
  });
});

describe("retries", () => {
  it("retries 5xx responses twice by default", async () => {
    const { client, calls } = mockClient([{ status: 500, body: {} }], {
      retryScheduleInMs: [1, 1],
    });

    await client.invoices.listInvoices().catch(() => null);

    assert.equal(calls.length, 3);
  });

  it("does not retry with numRetries: 0", async () => {
    const { client, calls } = mockClient([{ status: 500, body: {} }], {
      numRetries: 0,
    });

    const err = await client.invoices
      .listInvoices()
      .then(() => null)
      .catch((e: unknown) => e);

    assert.equal(calls.length, 1, "numRetries: 0 must produce exactly one attempt");
    assert.ok(err instanceof ApiException);
    assert.equal(err.status, 500);
  });

  it("does not retry with an empty retry schedule", async () => {
    const { client, calls } = mockClient([{ status: 500, body: {} }], {
      retryScheduleInMs: [],
    });

    await client.invoices.listInvoices().catch(() => null);

    assert.equal(calls.length, 1);
  });

  it("honours numRetries: 1", async () => {
    const { client, calls } = mockClient([{ status: 500, body: {} }], {
      numRetries: 1,
    });

    await client.invoices.listInvoices().catch(() => null);

    assert.equal(calls.length, 2);
    assert.equal(calls[1].init.headers["meteroid-retry-count"], "1");
  });

  it("does not retry 4xx responses", async () => {
    const { client, calls } = mockClient([
      { status: 400, body: { code: "x", detail: "y" } },
    ]);

    await client.invoices.listInvoices().catch(() => null);

    assert.equal(calls.length, 1);
  });
});

File: go/client_test.go
package meteroid

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"
	"time"
)

func testClient(t *testing.T, handler http.HandlerFunc) *Client {
	t.Helper()

	server := httptest.NewServer(handler)
	t.Cleanup(server.Close)

	return New("test-api-key", &Options{
		ServerURL: server.URL,
		// Keep the tests fast and deterministic.
		RetrySchedule: []time.Duration{},
	})
}

func TestListCustomers(t *testing.T) {
	const body = `{
		"data": [
			{
				"id": "cust_123",
				"name": "Test Customer",
				"currency": "USD",
				"custom_properties": {},
				"preferred_locales": [],
				"custom_taxes": [],
				"invoicing_emails": [],
				"invoicing_entity_id": "inv_1"
			}
		],
		"pagination_meta": {"page":0,"per_page":10,"total_items":1,"total_pages":1}
	}`

	var gotRequest *http.Request
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotRequest = r.Clone(r.Context())
		w.Header().Set("Content-Type", "application/json")
		io.WriteString(w, body)
	})

	response, err := client.Customers().ListCustomers(context.Background(), nil)
	if err != nil {
		t.Fatalf("ListCustomers: %v", err)
	}

	if gotRequest.Method != http.MethodGet {
		t.Errorf("method = %s, want GET", gotRequest.Method)
	}
	if gotRequest.URL.Path != "/api/v1/customers" {
		t.Errorf("path = %s", gotRequest.URL.Path)
	}
	if got := gotRequest.Header.Get("Authorization"); got != "Bearer test-api-key" {
		t.Errorf("authorization = %q", got)
	}
	if got := gotRequest.Header.Get("User-Agent"); got != "meteroid-go/"+Version {
		t.Errorf("user-agent = %q", got)
	}
	if got := gotRequest.Header.Get("Meteroid-Version"); got != APIVersion {
		t.Errorf("meteroid-version = %q", got)
	}

	if len(response.Data) != 1 {
		t.Fatalf("len(data) = %d, want 1", len(response.Data))
	}
	if response.Data[0].Name != "Test Customer" {
		t.Errorf("name = %q", response.Data[0].Name)
	}
	if response.PaginationMeta.TotalItems != 1 {
		t.Errorf("total_items = %d", response.PaginationMeta.TotalItems)
	}
}

// Optional query parameters are only sent when set, and repeated parameters are
// exploded rather than comma-joined.
func TestQueryParameterEncoding(t *testing.T) {
	var gotQuery string
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotQuery = r.URL.RawQuery
		io.WriteString(w, `{"data":[],"pagination_meta":{"page":0,"per_page":10,"total_items":0,"total_pages":0}}`)
	})

	page := int32(2)
	jobType := BatchJobType("INVOICE_ISSUE")
	_, err := client.BatchJobs().ListBatchJobs(context.Background(), &BatchJobsListBatchJobsOptions{
		Page:    &page,
		JobType: &jobType,
		Status:  []BatchJobStatus{BatchJobStatusPending, BatchJobStatusFailed},
	})
	if err != nil {
		t.Fatalf("ListBatchJobs: %v", err)
	}

	for _, want := range []string{"page=2", "job_type=INVOICE_ISSUE", "status=PENDING", "status=FAILED"} {
		if !strings.Contains(gotQuery, want) {
			t.Errorf("query %q is missing %q", gotQuery, want)
		}
	}
	if strings.Contains(gotQuery, "per_page") {
		t.Errorf("unset optional parameter was sent: %q", gotQuery)
	}
}

// Path parameters are substituted and escaped, POST bodies are sent as JSON, and
// an idempotency key is added automatically.
func TestCreateCustomer(t *testing.T) {
	var (
		gotPath string
		gotBody []byte
		gotKey  string
	)
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		gotBody, _ = io.ReadAll(r.Body)
		gotKey = r.Header.Get("idempotency-key")
		w.WriteHeader(http.StatusOK)
		io.WriteString(w, `{"id":"cust_1","name":"Acme","currency":"EUR","custom_properties":{},"custom_taxes":[],"invoicing_emails":[],"invoicing_entity_id":"inv_1","preferred_locales":[]}`)
	})

	customer, err := client.Customers().CreateCustomer(context.Background(), CustomerCreateRequest{
		Name:     Ptr("Acme"),
		Currency: CurrencyEur,
	})
	if err != nil {
		t.Fatalf("CreateCustomer: %v", err)
	}
	if customer.Id != "cust_1" {
		t.Errorf("id = %q", customer.Id)
	}
	if gotPath != "/api/v1/customers" {
		t.Errorf("path = %q", gotPath)
	}
	if !strings.HasPrefix(gotKey, "auto_") {
		t.Errorf("idempotency-key = %q, want an auto_ prefixed key", gotKey)
	}

	var sent map[string]any
	if err := json.Unmarshal(gotBody, &sent); err != nil {
		t.Fatalf("request body is not JSON: %s", gotBody)
	}
	if sent["name"] != "Acme" || sent["currency"] != "EUR" {
		t.Errorf("unexpected request body: %s", gotBody)
	}
	if _, ok := sent["alias"]; ok {
		t.Errorf("unset optional body field was sent: %s", gotBody)
	}
}

// A handful of OAuth endpoints take form-urlencoded bodies rather than JSON.
func TestFormEncodedBody(t *testing.T) {
	var (
		gotContentType string
		gotForm        url.Values
	)
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotContentType = r.Header.Get("Content-Type")
		if err := r.ParseForm(); err != nil {
			t.Errorf("ParseForm: %v", err)
		}
		gotForm = r.PostForm
		w.WriteHeader(http.StatusOK)
	})

	err := client.OAuth().RevokeEndpoint(context.Background(), RevocationRequest{
		Token:         "tok_123",
		TokenTypeHint: Ptr("access_token"),
	})
	if err != nil {
		t.Fatalf("RevokeEndpoint: %v", err)
	}

	if gotContentType != "application/x-www-form-urlencoded" {
		t.Errorf("content-type = %q", gotContentType)
	}
	if gotForm.Get("token") != "tok_123" {
		t.Errorf("token = %q", gotForm.Get("token"))
	}
	if gotForm.Get("token_type_hint") != "access_token" {
		t.Errorf("token_type_hint = %q", gotForm.Get("token_type_hint"))
	}
}

func TestPathParameterEscaping(t *testing.T) {
	var gotPath string
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		io.WriteString(w, `{"id":"cust_1","name":"Acme","currency":"EUR","custom_properties":{},"custom_taxes":[],"invoicing_emails":[],"invoicing_entity_id":"inv_1","preferred_locales":[]}`)
	})

	if _, err := client.Customers().GetCustomer(context.Background(), "alias with spaces/and-slash"); err != nil {
		t.Fatalf("GetCustomer: %v", err)
	}
	if gotPath != "/api/v1/customers/alias with spaces/and-slash" {
		t.Errorf("decoded path = %q", gotPath)
	}
}

func TestAPIErrorIsReturned(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		io.WriteString(w, `{"code":"NOT_FOUND","message":"no such customer"}`)
	})

	_, err := client.Customers().GetCustomer(context.Background(), "cust_missing")
	if err == nil {
		t.Fatal("expected an error")
	}

	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("error is %T, want *APIError", err)
	}
	if apiErr.StatusCode != http.StatusNotFound {
		t.Errorf("status = %d", apiErr.StatusCode)
	}
	want := RestErrorResponse{Code: ErrorCodeNotFound, Message: "no such customer"}
	if apiErr.Payload == nil || *apiErr.Payload != want {
		t.Errorf("payload = %+v, want %+v", apiErr.Payload, want)
	}
	if apiErr.OAuthPayload != nil {
		t.Errorf("oauth payload = %+v, want nil", apiErr.OAuthPayload)
	}
	if string(apiErr.RawBody) != `{"code":"NOT_FOUND","message":"no such customer"}` {
		t.Errorf("raw body = %q", apiErr.RawBody)
	}
	if got := apiErr.Error(); got != "meteroid: API error (status 404): no such customer (code NOT_FOUND)" {
		t.Errorf("message = %q", got)
	}
}

func TestOAuthErrorIsReturned(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		io.WriteString(w, `{"error":"invalid_grant"}`)
	})

	_, err := client.Customers().GetCustomer(context.Background(), "cust_123")

	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("error is %T, want *APIError", err)
	}
	if apiErr.StatusCode != http.StatusBadRequest {
		t.Errorf("status = %d", apiErr.StatusCode)
	}
	if apiErr.Payload != nil {
		t.Errorf("payload = %+v, want nil", apiErr.Payload)
	}
	if apiErr.OAuthPayload == nil || apiErr.OAuthPayload.Error != OAuthErrorCodeInvalidGrant {
		t.Fatalf("oauth payload = %+v", apiErr.OAuthPayload)
	}
	if apiErr.OAuthPayload.ErrorDescription != nil || apiErr.OAuthPayload.ErrorUri != nil {
		t.Errorf("optional fields should be nil: %+v", apiErr.OAuthPayload)
	}
	if !strings.Contains(apiErr.Error(), "invalid_grant") {
		t.Errorf("message = %q", apiErr.Error())
	}
}

func TestNonJSONErrorKeepsStatusAndBody(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadGateway)
		io.WriteString(w, "<html>Bad Gateway</html>")
	})

	_, err := client.Customers().GetCustomer(context.Background(), "cust_123")

	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("error is %T, want *APIError", err)
	}
	if apiErr.StatusCode != http.StatusBadGateway {
		t.Errorf("status = %d", apiErr.StatusCode)
	}
	if apiErr.Payload != nil || apiErr.OAuthPayload != nil {
		t.Errorf("expected no payload, got %+v / %+v", apiErr.Payload, apiErr.OAuthPayload)
	}
	if string(apiErr.RawBody) != "<html>Bad Gateway</html>" {
		t.Errorf("raw body = %q", apiErr.RawBody)
	}
	if !strings.Contains(apiErr.Error(), "Bad Gateway") {
		t.Errorf("message = %q", apiErr.Error())
	}
}

// A 422 goes through the same path as every other status: the API does not
// document one, so a body that is neither error schema has no payload.
func TestUnprocessableEntityUsesTheSamePath(t *testing.T) {
	const body = `{"detail":[{"loc":["body","name"],"msg":"field required","type":"missing"}]}`
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnprocessableEntity)
		io.WriteString(w, body)
	})

	_, err := client.Customers().CreateCustomer(context.Background(), CustomerCreateRequest{})

	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("error is %T, want *APIError", err)
	}
	if apiErr.StatusCode != http.StatusUnprocessableEntity {
		t.Errorf("status = %d", apiErr.StatusCode)
	}
	if apiErr.Payload != nil || apiErr.OAuthPayload != nil {
		t.Errorf("expected no payload, got %+v / %+v", apiErr.Payload, apiErr.OAuthPayload)
	}
	if string(apiErr.RawBody) != body {
		t.Errorf("raw body = %q", apiErr.RawBody)
	}
}

func TestNewAPIErrorDecoding(t *testing.T) {
	cases := []struct {
		name      string
		body      string
		wantRest  bool
		wantOAuth bool
	}{
		{"rest", `{"code":"CONFLICT","message":"exists"}`, true, false},
		// Error codes are open enums: an unknown code still decodes.
		{"rest unknown code", `{"code":"SOMETHING_NEW","message":"x"}`, true, false},
		{"rest missing message", `{"code":"CONFLICT"}`, false, false},
		{"rest null message", `{"code":"CONFLICT","message":null}`, false, false},
		{"rest wrong type", `{"code":"CONFLICT","message":42}`, false, false},
		{"oauth full", `{"error":"invalid_client","error_description":"bad secret","error_uri":null}`, false, true},
		{"empty object", `{}`, false, false},
		{"array", `[]`, false, false},
		{"empty body", ``, false, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			err := newAPIError(400, []byte(tc.body))
			if (err.Payload != nil) != tc.wantRest {
				t.Errorf("payload = %+v, want present=%v", err.Payload, tc.wantRest)
			}
			if (err.OAuthPayload != nil) != tc.wantOAuth {
				t.Errorf("oauth payload = %+v, want present=%v", err.OAuthPayload, tc.wantOAuth)
			}
			if string(err.RawBody) != tc.body {
				t.Errorf("raw body = %q", err.RawBody)
			}
		})
	}

	if err := newAPIError(400, []byte(`{"code":"SOMETHING_NEW","message":"x"}`)); err.Payload.Code.IsKnown() {
		t.Errorf("code %q should not be known", err.Payload.Code)
	}
}

// 5xx responses are retried; 4xx responses are not.
func TestRetries(t *testing.T) {
	var attempts int
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		attempts++
		if attempts < 3 {
			w.WriteHeader(http.StatusInternalServerError)
			io.WriteString(w, `{"code":"INTERNAL_SERVER_ERROR","message":"transient"}`)
			return
		}
		io.WriteString(w, `{"id":"cust_1","name":"Acme","currency":"EUR","custom_properties":{},"custom_taxes":[],"invoicing_emails":[],"invoicing_entity_id":"inv_1","preferred_locales":[]}`)
	}))
	defer server.Close()

	client := New("test-api-key", &Options{
		ServerURL:     server.URL,
		RetrySchedule: []time.Duration{time.Millisecond, time.Millisecond},
	})

	if _, err := client.Customers().GetCustomer(context.Background(), "cust_1"); err != nil {
		t.Fatalf("GetCustomer: %v", err)
	}
	if attempts != 3 {
		t.Errorf("attempts = %d, want 3", attempts)
	}
}

func TestClientErrorsAreNotRetried(t *testing.T) {
	var attempts int
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		attempts++
		w.WriteHeader(http.StatusBadRequest)
		io.WriteString(w, `{"code":"BAD_REQUEST","message":"nope"}`)
	})

	if _, err := client.Customers().GetCustomer(context.Background(), "cust_1"); err == nil {
		t.Fatal("expected an error")
	}
	if attempts != 1 {
		t.Errorf("attempts = %d, want 1", attempts)
	}
}

func TestContextCancellation(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		<-r.Context().Done()
	})

	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Millisecond)
	defer cancel()

	_, err := client.Customers().GetCustomer(ctx, "cust_1")
	if err == nil {
		t.Fatal("expected the request to fail once the context is cancelled")
	}
	var transportErr *TransportError
	if !errors.As(err, &transportErr) {
		t.Fatalf("error is %T (%v), want *TransportError", err, err)
	}
	if !errors.Is(err, context.DeadlineExceeded) {
		t.Errorf("errors.Is(err, context.DeadlineExceeded) = false for %v", err)
	}
	if transportErr.Method != http.MethodGet || transportErr.Path != "/api/v1/customers/{id_or_alias}" {
		t.Errorf("method/path = %s %s", transportErr.Method, transportErr.Path)
	}
}

// An explicit cancel, both during the round trip and while waiting between
// retries, surfaces as a *TransportError wrapping context.Canceled.
func TestExplicitCancellationIsATransportError(t *testing.T) {
	t.Run("in flight", func(t *testing.T) {
		ctx, cancel := context.WithCancel(context.Background())
		client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			cancel()
			<-r.Context().Done()
		})

		_, err := client.Customers().GetCustomer(ctx, "cust_1")
		var transportErr *TransportError
		if !errors.As(err, &transportErr) {
			t.Fatalf("error is %T (%v), want *TransportError", err, err)
		}
		if !errors.Is(err, context.Canceled) {
			t.Errorf("errors.Is(err, context.Canceled) = false for %v", err)
		}
	})

	t.Run("between retries", func(t *testing.T) {
		ctx, cancel := context.WithCancel(context.Background())
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusServiceUnavailable)
		}))
		defer server.Close()
		client := New("test-api-key", &Options{ServerURL: server.URL, RetrySchedule: []time.Duration{time.Hour}})

		// Cancel once the 503 is back and the client is waiting to retry.
		timer := time.AfterFunc(50*time.Millisecond, cancel)
		defer timer.Stop()

		_, err := client.Customers().GetCustomer(ctx, "cust_1")
		var transportErr *TransportError
		if !errors.As(err, &transportErr) {
			t.Fatalf("error is %T (%v), want *TransportError", err, err)
		}
		if !errors.Is(err, context.Canceled) {
			t.Errorf("errors.Is(err, context.Canceled) = false for %v", err)
		}
	})
}

// A server that cannot be reached is a *TransportError wrapping the net/http
// error, not an *APIError or a *DecodeError.
func TestUnreachableServerIsATransportError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	serverURL := server.URL
	server.Close()

	client := New("test-api-key", &Options{ServerURL: serverURL, RetrySchedule: []time.Duration{}})
	_, err := client.Customers().GetCustomer(context.Background(), "cust_1")

	var transportErr *TransportError
	if !errors.As(err, &transportErr) {
		t.Fatalf("error is %T (%v), want *TransportError", err, err)
	}
	var urlErr *url.Error
	if !errors.As(err, &urlErr) {
		t.Errorf("TransportError does not unwrap to the *url.Error: %v", err)
	}
	var apiErr *APIError
	var decodeErr *DecodeError
	if errors.As(err, &apiErr) || errors.As(err, &decodeErr) {
		t.Errorf("transport failure also matched APIError/DecodeError: %v", err)
	}
	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
		t.Errorf("connection refused reported as a context error: %v", err)
	}
}

// A 2xx response the SDK cannot decode is a *DecodeError carrying the status,
// the raw body and the json error.
func TestUndecodableResponseIsADecodeError(t *testing.T) {
	const body = `{"id":"cust_1","name":42}`
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusCreated)
		io.WriteString(w, body)
	})

	_, err := client.Customers().GetCustomer(context.Background(), "cust_1")

	var decodeErr *DecodeError
	if !errors.As(err, &decodeErr) {
		t.Fatalf("error is %T (%v), want *DecodeError", err, err)
	}
	if decodeErr.StatusCode != http.StatusCreated {
		t.Errorf("status = %d, want %d", decodeErr.StatusCode, http.StatusCreated)
	}
	if string(decodeErr.RawBody) != body {
		t.Errorf("raw body = %s, want %s", decodeErr.RawBody, body)
	}
	var typeErr *json.UnmarshalTypeError
	if !errors.As(err, &typeErr) {
		t.Errorf("DecodeError does not unwrap to the *json.UnmarshalTypeError: %v", err)
	}
	var apiErr *APIError
	var transportErr *TransportError
	if errors.As(err, &apiErr) || errors.As(err, &transportErr) {
		t.Errorf("decode failure also matched APIError/TransportError: %v", err)
	}

	// Malformed JSON too.
	client = testClient(t, func(w http.ResponseWriter, r *http.Request) {
		io.WriteString(w, `<html>not json</html>`)
	})
	_, err = client.Customers().GetCustomer(context.Background(), "cust_1")
	var syntaxErr *json.SyntaxError
	if !errors.As(err, &decodeErr) || decodeErr.StatusCode != http.StatusOK || !errors.As(err, &syntaxErr) {
		t.Errorf("malformed body: error is %T (%v), want *DecodeError wrapping *json.SyntaxError", err, err)
	}
}

func TestWithToken(t *testing.T) {
	var gotAuth string
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		io.WriteString(w, `{"id":"cust_1","name":"Acme","currency":"EUR","custom_properties":{},"custom_taxes":[],"invoicing_emails":[],"invoicing_entity_id":"inv_1","preferred_locales":[]}`)
	})

	if _, err := client.WithToken("other-key").Customers().GetCustomer(context.Background(), "cust_1"); err != nil {
		t.Fatalf("GetCustomer: %v", err)
	}
	if gotAuth != "Bearer other-key" {
		t.Errorf("authorization = %q", gotAuth)
	}
}
